doc/developer-guides/trusty.rst

```diff
@@ -0,0 +1,186 @@
+    |            |  | |        |     |        | | |
+    |    SOS     |  | | Normal |     | Secure | | |
+    |            |  | | World  |     | World  | |-+
+    |            |  | |        |     |        | |
+    |            |  | +--------+     +--------+ |
+    +------------+  +---------------------------+
+    +-------------------------------------------+
+    |             ACRN Hypervisor               |
+    +-------------------------------------------+
+    +-------------------------------------------+
+    |                   HW                      |
+    +-------------------------------------------+
+
+.. note::
+   Trusty OS is running in Secure World in the architecture above.
+
+Trusty specific Hypercalls
+**************************
+
+1. ``HC_INITIALIZE_TRUSTY``
+
+   * This Hypercall is used by the User OS bootloader (``UOS_Loader``) to request ACRN to initialize Trusty
+   * The Trusty memory region range, entry point must be specified
+   * Hypervisor needs to save current vCPU contexts (Normal World)
+#. ``HC_WORLD_SWITCH``
+
+   * Simulate ARM SMC (Secure Monitor Call, or SMC) instruction to do world switch
+   * Hypervisor needs to save current world vCPU contexts, and load next world vCPU contexts
+   * Update rdi, rsi, rdx, rbx to next world vCPU contexts
+
+APIs
+====
+
+.. code-block:: c
+
+   hcall_initialize_trusty(vm_t *vm)
+   hcall_world_switch(vm_t *vm)
+
+
+Trusty Boot flow
+****************
+
+By design, the User OS bootloader (``UOS_Loader``) will trigger the Trusty boot process. The complete bootflow is illustrated below.
+
+.. graphviz:: trusty/boot-flow.dot
+   :name: trusty-boot-flow
+   :align: center
+   :caption: Trusty Boot Flow
+
+Details
+
+1. UOS_Loader
+
+   a. Load and verify trusty image from virtual disk
+   #. Allocate runtime memory for trusty
+   #. Do ELF relocation of trusty image and get entry address
+   #. Call ``HC_INITIALIZE_TRUSTY`` with trusty memory base and entry address
+#. ACRN (``HC_INITIALIZE_TRUSTY``)
+
+   a. Save World context for Normal World
+   #. Init World context for Secure World(RIP, RSP, EPT, etc.)
+   #. Resume to Secure World
+#. Trusty
+
+   a. Booting up
+   #. Call ``HC_WORLD_SWITCH`` to switch back to Normal World if boot completed
+#. ACRN (``HC_WORLD_SWITCH``)
+
+   a. Save World context for the World which caused this ``vmexit`` (Secure World)
+   #. Restore World context for next World (Normal World (UOS_Loader))
+   #. Resume to next World (UOS_Loader)
+#. UOS_Loader
+
+   a. Continue to boot
+
+EPT Hierarchy
+*************
+
+As per the Trusty design, Trusty can access Normal World's memory, but Normal World cannot
+access Secure World's memory. Hence it means Secure World EPTP page table hierarchy
+must contain normal world GPA address space, while Trusty world's GPA address space
+must be removed from the Normal world EPTP page table hierarchy.
+
+Design
+======
+
+Put Secure World's GPA to very high position: 511G-512G. The PML4/PDPT for Trusty
```

Comment:
We should use the abbreviation ``GB`` for Gigabytes rather than just ``G``, and there should be a space between the number and the units, ``511 GB - 512 GB``

Reply:
OK